Implement CSS3 Calc

dzbarsky opened the pull request:
Doesn't work for font-size and I need to investigate how it should work for units that aren't lengths, but what I have so far should be ok. r? @SimonSapin

Fixes #7156.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/7185)

<!-- Reviewable:end -->

dzbarsky committed:
Implement Calc for LengthOrPercentage

dzbarsky committed:
Implement Calc for LengthOrPercentageOrAuto

dzbarsky committed:
Implement proper calc parsing

dzbarsky committed:
Implement font relative and viewport relative units for calc

dzbarsky committed:
Clean up serialization and other hacks

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify layout code, but no reftests are modified. Please consider adding a reftest!

dzbarsky commented:
Also need to figure out if we have tests or can steal some from Gecko.

dzbarsky commented:
Pretty sure the AST simplification code isn't quite right and can actually enter some places I thought were unreachable.  I'll plug those up as part of writing the tests, but let me know if you can think of a cleaner way to do this.

dzbarsky committed:
Clean up serialization code a little

dzbarsky committed:
Clean up AST simplification code

pcwalton commented on components/style/values.rs:
```diff
@@ -353,11 +353,317 @@ pub mod specified {
+            let ast = try!(Calc::parse_sum(input));
+            let ast = try!(Calc::simplify_sum(ast));
+
+            println!("Simlified ast {:?} ", ast);
```
(Remove this before committing)

dzbarsky committed:
Properly serialize % values in calc expressions

dzbarsky committed:
Simplify like terms in all sum expressions, not just toplevel calc

dzbarsky committed:
Simplify the calc AST simplification code

dzbarsky committed:
Remove stray changes

dzbarsky committed:
Fix some calc parsing panics

dzbarsky commented:
Ok, the implementation now passes all the tests I've written.  I want to figure out if I can somehow clean up the code because it's a little messy

dzbarsky committed:
Expand out nested products

dzbarsky committed:
Use the type system to enforce that product nodes are simplified away

dzbarsky committed:
Get rid of some cloning

SimonSapin commented:
@dzbarsky Do you want to keep iterating on this some more, or is this ready for review?

dzbarsky commented:
It's ready for review now. I'm gone for the next week so there's no rush.

dzbarsky committed:
Clean up some stray lines

SimonSapin commented:
Reviewed 1 of 6 files at r1, 5 of 7 files at r2, 1 of 2 files at r3, 3 of 4 files at r8, 1 of 2 files at r11.
Review status: 11 of 12 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

<sup>**[components/layout/block.rs, line 333 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwUINQjoCBD3FFZTVN)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/layout/block.rs#L333)):</sup>
Is this ignoring the percentage component? I think this should rather resolved to `auto`: the "real" value that we would want is not known yet, same as the percentage case in https://drafts.csswg.org/css2/visudet.html#the-height-property

Though perhaps there should be a special case when the percentage component is zero.

---

<sup>**[components/layout/block.rs, line 334 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwV9FYV2Hj105oKalD)** ([raw file](https://github.com/servo/servo/blob/84c4a26e4d05015da4cbbf1f120b317a23bec0b4/components/layout/block.rs#L334)):</sup>
Nit: I think a `{}` block is more readable when you’re going to the next line anyway… hopefully rustfmt will make this a non-issue eventually.

---

<sup>**[components/layout/block.rs, line 347 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwVULq0aerDu3rjHTX)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/layout/block.rs#L347)):</sup>
Same comment as above regarding ignoring the percentage component.

---

<sup>**[components/layout/block.rs, line 1130 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwVYuIta4vTibBUo5D)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/layout/block.rs#L1130)):</sup>
Same comment as above regarding ignoring the percentage component.

---

<sup>**[components/layout/fragment.rs, line 459 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwWPlV08RLXC0C88Xy)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/layout/fragment.rs#L459)):</sup>
Same comment as above regarding ignoring the percentage component.

---

<sup>**[components/layout/fragment.rs, line 615 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwW4YjRDOH9XPmpvCI)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/layout/fragment.rs#L615)):</sup>
Nit: This sum is duplicated all over the place. Maybe make it a method of `Calc` or even of `LengthOrPercentage`?

---

<sup>**[components/style/properties.mako.rs, line 1864 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwXJ_4Gcv_a4LtBRdb)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/style/properties.mako.rs#L1864)):</sup>
Reject unsupported things by returning `Err`, don’t make up an incorrect value.

---

<sup>**[components/style/values.rs, line 423 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwYrYfCIGUQ0NdhnTh)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/style/values.rs#L423)):</sup>
Is `None` different from `Some(zero)`?

---

<sup>**[components/style/values.rs, line 1287 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwZP6WBNx_Wn1zBFJL)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/style/values.rs#L1287)):</sup>
Is `None` different from `Some(zero)`?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7185)
<!-- Sent from Reviewable.io -->

dzbarsky commented:
Review status: 11 of 12 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

<sup>**[components/layout/block.rs, line 333 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwUINQjoCBD3FFZTVN)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/layout/block.rs#L333)):</sup>
Hmmm, I think you're probably right, but I think that will end up ignoring the absolute component. But maybe that's ok. I'll investigate when I write reftests.

---

<sup>**[components/layout/block.rs, line 334 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwV9FYV2Hj105oKalD)** ([raw file](https://github.com/servo/servo/blob/84c4a26e4d05015da4cbbf1f120b317a23bec0b4/components/layout/block.rs#L334)):</sup>
Ok I'll put it back.

---

<sup>**[components/layout/block.rs, line 347 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwVULq0aerDu3rjHTX)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/layout/block.rs#L347)):</sup>
Percentages without a containing block size are given a min_block_size of 0 here, so it seems ok to ignore the percentage and just use the absolute length (when there is no containing block size).

---

<sup>**[components/layout/block.rs, line 1130 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwVYuIta4vTibBUo5D)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/layout/block.rs#L1130)):</sup>
Same as above, percentage returns None or a fraction of containing block size, so calc is doing the same thing.

---

<sup>**[components/layout/fragment.rs, line 0 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwW4YjRDOH9XPmpvCI)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/layout/fragment.rs#L0)):</sup>
Yeah, I think it makes sense on Calc, since we'll also need it on LengthOrPercentageOrAuto

---

<sup>**[components/layout/fragment.rs, line 459 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwWPlV08RLXC0C88Xy)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/layout/fragment.rs#L459)):</sup>
Agreed, this should probably be Auto.

---

<sup>**[components/style/properties.mako.rs, line 1864 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwXJ_4Gcv_a4LtBRdb)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/style/properties.mako.rs#L1864)):</sup>
Good point, I was worried about throwing but we were already throwing for this case

---

<sup>**[components/style/values.rs, line 423 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwYrYfCIGUQ0NdhnTh)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/style/values.rs#L423)):</sup>
Yes. https://drafts.csswg.org/css-values-3/#calc-computed-value "Where percentages are not resolved at computed-value time, they are not resolved in calc() expressions, e.g. calc(100% - 100% + 1em) resolves to calc(0% + 1em), not to calc(1em)." So Some(zero) isn't enough to distinguish these two cases.

---

<sup>**[components/style/values.rs, line 1287 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwZP6WBNx_Wn1zBFJL)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/style/values.rs#L1287)):</sup>
Yes. calc(100px) serializes as 100px, while calc(100px + 0%) stays as a calc expression. "A value like calc(20px + 0%) would serialize as calc(20px + 0%), maintaining both terms in the serialized value. (It’s important to maintain zero-valued terms, so the calc() doesn’t suddenly "change shape" in the middle of a transition when one of the values happens to have a zero value temporarily. This also removes the need to "pick a unit" when all the terms are zero.)"

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7185)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7265) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin commented:
Reviewed 1 of 1 files at r17.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

<sup>**[components/layout/block.rs, line 333 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwUINQjoCBD3FFZTVN)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/layout/block.rs#L333)):</sup>
The reasoning is that the sum of something unknown plus something known is also unknown.

---

<sup>**[components/layout/block.rs, line 347 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwVULq0aerDu3rjHTX)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/layout/block.rs#L347)):</sup>
Alright, makes sense. (x + 0 = x)

---

<sup>**[components/layout/block.rs, line 1130 \[r16\]](https://reviewable.io:443/reviews/servo/servo/7185#-JwwVYuIta4vTibBUo5D)** ([raw file](https://github.com/servo/servo/blob/e6620fba78c3da05f3219d0feb4e3823d94fd532/components/layout/block.rs#L1130)):</sup>
But `None` is more like "unknown" than like zero. Is this is more similar to the first case: a known + unknown sum is unknown.

---

<sup>**[components/style/values.rs, line 442 \[r17\]](https://reviewable.io:443/reviews/servo/servo/7185#-JxFsXQmMBIM2LAr7qyw)** ([raw file](https://github.com/servo/servo/blob/e85c005e7ca1a497ad28e93b870335f86806bee1/components/style/values.rs#L442)):</sup>
https://github.com/servo/servo/issues/7306 can be a follow-up later.

---

<sup>**[components/style/values.rs, line 1303 \[r17\]](https://reviewable.io:443/reviews/servo/servo/7185#-JxGdRSwyHum1-ZUBmAb)** ([raw file](https://github.com/servo/servo/blob/e85c005e7ca1a497ad28e93b870335f86806bee1/components/style/values.rs#L1303)):</sup>
Is this missing `* 100.`?

---

<sup>**[components/style/values.rs, line 1327 \[r17\]](https://reviewable.io:443/reviews/servo/servo/7185#-JxGdoPcg2wZO248MweL)** ([raw file](https://github.com/servo/servo/blob/e85c005e7ca1a497ad28e93b870335f86806bee1/components/style/values.rs#L1327)):</sup>
Use `&[] ` instead of `vec!()` here and in the previous loop to avoid allocating.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7185)
<!-- Sent from Reviewable.io -->

dzbarsky committed:
Address review comments

dzbarsky commented:
Review status: 1 of 17 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[components/style/values.rs, line 442 \[r17\]](https://reviewable.io:443/reviews/servo/servo/7185#-JxFsXQmMBIM2LAr7qyw)** ([raw file](https://github.com/servo/servo/blob/e85c005e7ca1a497ad28e93b870335f86806bee1/components/style/values.rs#L442)):</sup>
I pushed a new commit with tests for this - I don't think #7306 is actually an issue.

---

<sup>**[components/style/values.rs, line 1303 \[r17\]](https://reviewable.io:443/reviews/servo/servo/7185#-JxGdRSwyHum1-ZUBmAb)** ([raw file](https://github.com/servo/servo/blob/e85c005e7ca1a497ad28e93b870335f86806bee1/components/style/values.rs#L1303)):</sup>
Yep!

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7185)
<!-- Sent from Reviewable.io -->

dzbarsky committed:
Add calc reftest

dzbarsky committed:
Add calc parsing tests for the other properties

jdm commented:
Is this superseded by #7400?

dzbarsky commented:
Yes, but I can either rebase #7400 after this merges or merge it all at once. Whichever is easier for @SimonSapin

SimonSapin commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 2591a89 has been approved by `SimonSapin`

<!-- @bors-servo r=SimonSapin 2591a89a91945629cee5ea5df8598eff5334f24e -->

bors-servo commented:
:hourglass: Testing commit 2591a89 with merge 8f39e46...

bors-servo commented:
:broken_heart: Test failed - [gonk](http://build.servo.org/builders/gonk/builds/2111)

SimonSapin commented:
```rust
/home/servo/buildbot/slave/gonk/build/components/style/values.rs:609:25: 616:26 error: non-exhaustive patterns: `Ch(_)` not covered [E0004]
/home/servo/buildbot/slave/gonk/build/components/style/values.rs:609                         match f {
/home/servo/buildbot/slave/gonk/build/components/style/values.rs:610                             FontRelativeLength::Em(val) =>
/home/servo/buildbot/slave/gonk/build/components/style/values.rs:611                                 em = Some(em.unwrap_or(0.) + val),
/home/servo/buildbot/slave/gonk/build/components/style/values.rs:612                             FontRelativeLength::Ex(val) =>
/home/servo/buildbot/slave/gonk/build/components/style/values.rs:613                                 ex = Some(ex.unwrap_or(0.) + val),
/home/servo/buildbot/slave/gonk/build/components/style/values.rs:614                             FontRelativeLength::Rem(val) =>
                                                                     ...
note: in expansion of for loop expansion
/home/servo/buildbot/slave/gonk/build/components/style/values.rs:591:13: 620:14 note: expansion site

nox commented:
Isn't LengthOrPercentageOrAuto a misnomer now?

SimonSapin commented:
Indeed, though the grammar in specs is still `<length> | <percentage> | auto`. You can combine lengths with percentages, but not `auto` with anything else. Do you have ideas for another name? (And for other similar types.)

nox commented:
Oh I see, I guess that's not a misnomer then. Or at least LengthOrPercentage isn't.

SimonSapin commented:
See also #7496.

dzbarsky committed:
Add ch support

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7510) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin commented:
I had hoped this would auto-merge with #7496. Is there something here that’s not there?

dzbarsky commented:
Yeah, I added a commit adding ch support and modified the test. I see you merged your version already so I'll just roll the test change into the other calc PR I have open.
